draft-ietf-jsonpath-base.md

```diff
@@ -413,51 +433,70 @@ the grammar in this document, its UTF-8 form SHOULD first be decoded into
-sometimes referred to as the type of the node.
-
-
-## Implementation
-
-An implementation of this specification, from now on referred to simply as
-"an implementation", SHOULD takes two inputs, a JSONPath and a JSON value,
-and produce
-a possibly empty list of nodes of the JSON value which are selected by
-the JSONPath or an error (but not both).
-
-If no node is selected and no error has occurred, an implementation MUST
-return an empty list of nodes.
-
-Syntax errors in the JSONPath SHOULD be detected before selection is attempted
-since these errors do not depend on the JSON value.
-Therefore, an implementation SHOULD take a JSONPath and produce an optional
-syntax error and then,
-if and only if an error was not produced, SHOULD take a JSON value and
-produce a list of nodes or an error (but not both).
-
-Alternatively, an implementation MAY take a JSONPath and a JSON value
-and produce a list of nodes or an optional error (but not both).
-
-For any implementation, if a syntactically invalid JSONPath is provided,
-the implementation MUST return an error.
-
-If a syntactially invalid JSON value is provided, any implementation SHOULD
-return an error.
+A string to be used as a JSONPath query needs to be *well-formed* and
+*valid*.
+A string is a well-formed JSONPath query if it conforms to the syntax
+of JSONPath.
+A well-formed JSONPath query is valid if it also fulfills all semantic
+requirements posed by this document.
+
+Well-formedness and validity of JSONPath queries are independent of
+the JSON value the query is applied to; no further errors can be
+raised during application of the query to a JSON value.
+
+(Obviously, an implementation can still fail when executing a JSONPath
+query, e.g., because of resource depletion, but this is not modeled in
+the present specification.)
+
+> Discussion: are we ready for the error model defined here, i.e. one
+> that does not have runtime errors, but only compile time errors?
+
+## Processing Model
+
+A valid JSONPath query is executed against a JSON value, the *Argument*, and
+produces a list of zero or more nodes of the JSON value which are
+selected by the JSONPath.
+(Note that the term JSON value also implies that this value complies
+to the definitions in {{-json}}, i.e., is indeed a JSON value.)
+
+The JSONPath query is a sequence of zero or more *selectors*, each of
+which is applied to the result of the previous selector and leaves
+input to the next selector.
+These results and inputs take the form of a *nodelist*, i.e., a
+sequence of zero or more nodes.
+
+The nodelist going into the first selector contains a single node,
+holding the Argument.
+The nodelist resulting from the last selector is presented as the
+result of the query; depending on the specific API, it might be
+presented as an array of the JSON values at the nodes, an array of
+Output Paths referencing the nodes, or both — or some other
+representation as desired by the implementation.
```

Comment:
I like leaving this open to the implementation.  We _could_ (in a later PR) go so far as to define what the output looks like for each of these cases in order to give an expected API for consumers of the implementations.

It may also be beneficial to explicitly state that a node list is just a collection of nodes; it is _not_ a JSON array, although its _representation in JSON_ is.  This is to say that a node list can be _serialized_ to a JSON array (e.g. for the purposes of output), but it is not one inherently.

Reply:
Right.  The current PR text says "depending on the specific API", so I think we are covered on the first item.  It also doesn't seem to suggest the nodelist is a JSON array, or are text changes actually required for the second suggestion?

Reply:
I don't think that we should rely on implications.  Explicit statements are always better.

Reply:
Can you say what the wording should be?

Reply:
Actually, it looks like lines 221-222 cover it: 

> While this list can be represented in JSON, e.g. as an array, the nodelist is an abstract concept unrelated to JSON data items.